# Code Example

In [2]:
import pandas as pd
import numpy as np


In [3]:
df = pd.read_csv('WorldPopulation2022.csv')
df.head()

# Country (or dependency) Population (2022) Yearly change  Net change  \
0  1                   China     1,439,323,776         0.39%   5,540,090   
1  2                   India     1,380,004,385         0.99%  13,586,631   
2  3           United States       331,002,651         0.59%   1,937,734   
3  4               Indonesia       273,523,615         1.07%   2,898,047   
4  5                Pakistan       220,892,340         2.00%   4,327,022   

  Density (P/Km²) Land Area (Km²) Migrants (net) Fert. Rate Med.Age  \
0             153       9,388,211       -348,399        1.7      38   
1             464       2,973,190       -532,687        2.2      28   
2              36       9,147,420        954,806        1.8      38   
3             151       1,811,570        -98,955        2.3      30   
4             287         770,880       -233,379        3.6      23   

  Urban Pop % World Share  
0         61%      18.47%  
1         35%      17.70%  
2         83%       4.25%  
3         56%       3.51%  
4         35%       2.83%

In [4]:
df.shape

(235, 12)

There are 235 rows and 12 columns (features)

## Exploratory Data Analysis

In [5]:
df = df.drop('#', axis=1)

In [7]:
df.head()

Country (or dependency) Population (2022) Yearly change  Net change  \
0                   China     1,439,323,776         0.39%   5,540,090   
1                   India     1,380,004,385         0.99%  13,586,631   
2           United States       331,002,651         0.59%   1,937,734   
3               Indonesia       273,523,615         1.07%   2,898,047   
4                Pakistan       220,892,340         2.00%   4,327,022   

  Density (P/Km²) Land Area (Km²) Migrants (net) Fert. Rate Med.Age  \
0             153       9,388,211       -348,399        1.7      38   
1             464       2,973,190       -532,687        2.2      28   
2              36       9,147,420        954,806        1.8      38   
3             151       1,811,570        -98,955        2.3      30   
4             287         770,880       -233,379        3.6      23   

  Urban Pop % World Share  
0         61%      18.47%  
1         35%      17.70%  
2         83%       4.25%  
3         56%       3.51%  
4         35%       2.83%

In [9]:
df = df.rename(columns={'Country (or dependency)':'Country', 'Population (2022)':'Population', 'Med.Age': 'Median Age', 'Fert. Rate':'Fertility Rate', 'Migrants (net)': 'Migrants'})
df.head()

Country     Population Yearly change  Net change Density (P/Km²)  \
0          China  1,439,323,776         0.39%   5,540,090             153   
1          India  1,380,004,385         0.99%  13,586,631             464   
2  United States    331,002,651         0.59%   1,937,734              36   
3      Indonesia    273,523,615         1.07%   2,898,047             151   
4       Pakistan    220,892,340         2.00%   4,327,022             287   

  Land Area (Km²)  Migrants Fertility Rate MedianAge Urban Pop % World Share  
0       9,388,211  -348,399            1.7        38         61%      18.47%  
1       2,973,190  -532,687            2.2        28         35%      17.70%  
2       9,147,420   954,806            1.8        38         83%       4.25%  
3       1,811,570   -98,955            2.3        30         56%       3.51%  
4         770,880  -233,379            3.6        23         35%       2.83%

In [12]:
df.dtypes

Country            object
Population         object
Yearly change      object
Net change         object
Density (P/Km²)    object
Land Area (Km²)    object
Migrants           object
Fertility Rate     object
MedianAge          object
Urban Pop %        object
World Share        object
dtype: object

Not good, we need all these to be of type int (execpt for country name)

In [29]:
df.loc[:,'Population':] = df.loc[:,'Population':].astype(float)

ValueError: could not convert string to float: '1,439,323,776'

In [30]:
df.loc[:,'Population':].apply()

Population Yearly change  Net change Density (P/Km²) Land Area (Km²)  \
0    1,439,323,776         0.39%   5,540,090             153       9,388,211   
1    1,380,004,385         0.99%  13,586,631             464       2,973,190   
2      331,002,651         0.59%   1,937,734              36       9,147,420   
3      273,523,615         1.07%   2,898,047             151       1,811,570   
4      220,892,340         2.00%   4,327,022             287         770,880   
..             ...           ...         ...             ...             ...   
230          4,992         0.06%           3              50             100   
231          3,480         3.05%         103               0          12,170   
232          1,626         0.68%          11               6             260   
233          1,357         1.27%          17             136              10   
234            801         0.25%           2           2,003               0   

     Migrants Fertility Rate MedianAge Urban Pop % World Share  
0    -348,399            1.7        38         61%      18.47%  
1    -532,687            2.2        28         35%      17.70%  
2     954,806            1.8        38         83%       4.25%  
3     -98,955            2.3        30         56%       3.51%  
4    -233,379            3.6        23         35%       2.83%  
..        ...            ...       ...         ...         ...  
230       NaN           N.A.      N.A.         10%       0.00%  
231       NaN           N.A.      N.A.         66%       0.00%  
232       NaN           N.A.      N.A.         46%       0.00%  
233       NaN           N.A.      N.A.          0%       0.00%  
234       NaN           N.A.      N.A.        N.A.       0.00%  

[235 rows x 10 columns]

In [11]:
df.isnull().sum()

Country             0
Population          0
Yearly change       0
Net change          0
Density (P/Km²)     0
Land Area (Km²)     0
Migrants           34
Fertility Rate      0
MedianAge           0
Urban Pop %         0
World Share         0
dtype: int64

229,Saint Helena,"6,077",0.30%,18,16,390,,N.A.,N.A.,27%,0.00%

I am unsure of why they decided to use N.A. as an na category type
when both pandas and numpy have na specific types

either pd.NA or np.nan
* the np.nan is actually of type float

